src/Verilator.cpp

```diff
@@ -526,6 +526,54 @@ static void process() {
     // Note early return above when opt.cdc()
 }
 
+static void execBuildJob() {
+    if (!v3Global.opt.build()) return;
+
+    std::stringstream cmd;
+    const V3StringList& makeFlags = v3Global.opt.makeFlags();
+    const int jobs = v3Global.opt.buildJobs();
+    if (v3Global.opt.gmake()) {  // If both gmake and cmake are chosen, use gmake to build.
+        cmd <<  v3Global.opt.makeCmd();
+        cmd << " -C " << v3Global.opt.makeDir();
+        cmd << " -f " << v3Global.opt.prefix() << ".mk";
+        if (jobs == 0) {
+            cmd << " -j";
+        } else if (jobs > 1) {
+            cmd << " -j " << jobs;
+        }
+        for (V3StringList::const_iterator it = makeFlags.begin(); it != makeFlags.end(); ++it) {
+            cmd << ' ' << *it;
+        }
+    } else {
+        UASSERT(v3Global.opt.cmake(), "at least cmake or gmake must be chosen in V3Options.cpp");
+        cmd << "cd " << v3Global.opt.makeDir() << " && ";
+        cmd << "cmake";
+        for (V3StringList::const_iterator it = makeFlags.begin(); it != makeFlags.end(); ++it) {
+            cmd << ' ' << *it;
+        }
+        cmd << ' ' << V3Os::getcwd() << " && ";
```

Comment:
I think we should use relative paths (never getcwd) because using an absolute path can cause the objects to refer to absolute filenames, which makes ccache not be able to cache hit between multiple directories/users.

Reply:
Hmm, I hesitate to add such change because of the following 3 reasons.

1. cmake looks to specify .cpp to compiler using absolute path even though relative path is passed to cmake
2. ccache has an [option](https://ccache.dev/manual/latest.html#_compiling_in_different_directories) of `base_dir` for the usecase you mentioned.
3.  getting a relative path is not trivial when considering symlink. [C++17](https://en.cppreference.com/w/cpp/filesystem/relative) and [boost](https://www.boost.org/doc/libs/1_72_0/libs/filesystem/doc/reference.html#op-relative) provides though.

Because I am a casual ccache user and know little about cmake, I did a small experiment to see how cmake specify .cpp to g++.
## Small Experiment
### Files

CMakeLists.txt looks as below. ( I think this is minimal example.)
```cmake
cmake_minimum_required(VERSION 3.14)
project(cmake_example)
find_package(verilator HINTS $ENV{VERILATOR_ROOT})
include_directories(obj_dir)
add_executable(Vt src/main.cpp)
verilate(Vt SOURCES src/t.v)
```
There are just `main.cpp` and `t.v` in the directory `src`.
### Run with the relative path
#### Command
I ran the following shell script.
```bash
#!/bin/bash
set -xe
export VERILATOR_ROOT=$HOME/prj/verilator
rm -rf ./obj_dir
${VERILATOR_ROOT}/bin/verilator  -Wno-fatal --cc --exe --top-module t --make cmake src/t.v src/main.cpp 
cd obj_dir
cmake ..
cmake --build . -v
```
Note that `verilator` has no `--build` option. Both `cmake` for configuration and `cmake --build` are launched with the relative path.

#### Build commands
Here is a log for build.  Apparently all .cpp files are specified by the absolute path.
```
+ cmake --build . -v
/usr/bin/cmake -S/home/yutetsu/prj/verilator_cases/11 -B/home/yutetsu/prj/verilator_cases/11/obj_dir --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles /home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/progress.marks
/usr/bin/gmake -f CMakeFiles/Makefile2 all
gmake[1]: Entering directory '/mnt/prj/verilator_cases/11/obj_dir'
/usr/bin/gmake -f CMakeFiles/Vt.dir/build.make CMakeFiles/Vt.dir/depend
gmake[2]: Entering directory '/mnt/prj/verilator_cases/11/obj_dir'
cd /home/yutetsu/prj/verilator_cases/11/obj_dir && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/yutetsu/prj/verilator_cases/11 /home/yutetsu/prj/verilator_cases/11 /home/yutetsu/prj/verilator_cases/11/obj_dir /home/yutetsu/prj/verilator_cases/11/obj_dir /home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/DependInfo.cmake --color=
Dependee "/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/DependInfo.cmake" is newer than depender "/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/depend.internal".
Dependee "/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/depend.internal".
Scanning dependencies of target Vt
gmake[2]: Leaving directory '/mnt/prj/verilator_cases/11/obj_dir'
/usr/bin/gmake -f CMakeFiles/Vt.dir/build.make CMakeFiles/Vt.dir/build
gmake[2]: Entering directory '/mnt/prj/verilator_cases/11/obj_dir'
[ 14%] Building CXX object CMakeFiles/Vt.dir/src/main.cpp.o
/usr/bin/c++  -DVM_COVERAGE=0 -DVM_SC=0 -DVM_TRACE=0 -DVM_TRACE_FST=0 -DVM_TRACE_VCD=0 -I/home/yutetsu/prj/verilator_cases/11/obj_dir -I/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/Vt.dir -I/home/yutetsu/prj/verilator/include -I/home/yutetsu/prj/verilator/include/vltstd   -o CMakeFiles/Vt.dir/src/main.cpp.o -c /home/yutetsu/prj/verilator_cases/11/src/main.cpp
[ 28%] Building CXX object CMakeFiles/Vt.dir/CMakeFiles/Vt.dir/Vt.dir/Vt.cpp.o
/usr/bin/c++  -DVM_COVERAGE=0 -DVM_SC=0 -DVM_TRACE=0 -DVM_TRACE_FST=0 -DVM_TRACE_VCD=0 -I/home/yutetsu/prj/verilator_cases/11/obj_dir -I/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/Vt.dir -I/home/yutetsu/prj/verilator/include -I/home/yutetsu/prj/verilator/include/vltstd   -o CMakeFiles/Vt.dir/CMakeFiles/Vt.dir/Vt.dir/Vt.cpp.o -c /home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/Vt.dir/Vt.cpp
[ 42%] Building CXX object CMakeFiles/Vt.dir/CMakeFiles/Vt.dir/Vt.dir/Vt__Syms.cpp.o
/usr/bin/c++  -DVM_COVERAGE=0 -DVM_SC=0 -DVM_TRACE=0 -DVM_TRACE_FST=0 -DVM_TRACE_VCD=0 -I/home/yutetsu/prj/verilator_cases/11/obj_dir -I/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/Vt.dir -I/home/yutetsu/prj/verilator/include -I/home/yutetsu/prj/verilator/include/vltstd   -o CMakeFiles/Vt.dir/CMakeFiles/Vt.dir/Vt.dir/Vt__Syms.cpp.o -c /home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/Vt.dir/Vt__Syms.cpp
[ 57%] Building CXX object CMakeFiles/Vt.dir/home/yutetsu/prj/verilator/include/verilated.cpp.o
/usr/bin/c++  -DVM_COVERAGE=0 -DVM_SC=0 -DVM_TRACE=0 -DVM_TRACE_FST=0 -DVM_TRACE_VCD=0 -I/home/yutetsu/prj/verilator_cases/11/obj_dir -I/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/Vt.dir -I/home/yutetsu/prj/verilator/include -I/home/yutetsu/prj/verilator/include/vltstd   -o CMakeFiles/Vt.dir/home/yutetsu/prj/verilator/include/verilated.cpp.o -c /home/yutetsu/prj/verilator/include/verilated.cpp
[ 71%] Linking CXX executable Vt
/usr/bin/cmake -E cmake_link_script CMakeFiles/Vt.dir/link.txt --verbose=1
/usr/bin/c++     CMakeFiles/Vt.dir/src/main.cpp.o CMakeFiles/Vt.dir/CMakeFiles/Vt.dir/Vt.dir/Vt.cpp.o CMakeFiles/Vt.dir/CMakeFiles/Vt.dir/Vt.dir/Vt__Syms.cpp.o CMakeFiles/Vt.dir/home/yutetsu/prj/verilator/include/verilated.cpp.o  -o Vt 
gmake[2]: Leaving directory '/mnt/prj/verilator_cases/11/obj_dir'
[100%] Built target Vt
gmake[1]: Leaving directory '/mnt/prj/verilator_cases/11/obj_dir'
/usr/bin/cmake -E cmake_progress_start /home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles 0
```

### Run with the absolute path
#### Command
Absolute path is passed to cmake this time.
```bash
#!/bin/bash
set -xe
export VERILATOR_ROOT=$HOME/prj/verilator
rm -rf ./obj_dir
${VERILATOR_ROOT}/bin/verilator  -Wno-fatal --cc --exe --top-module t --make cmake src/t.v src/main.cpp 
cd obj_dir
cmake /home/yutetsu/prj/verilator_cases/11
cmake --build /home/yutetsu/prj/verilator_cases/11/obj_dir -v
```
#### Build commands
Here is a log for build.
No difference is observed.
- .cpp is specified by the absolute path
- .o is specified by the relative path
```
+ cmake --build /home/yutetsu/prj/verilator_cases/11/obj_dir -v
/usr/bin/cmake -S/home/yutetsu/prj/verilator_cases/11 -B/home/yutetsu/prj/verilator_cases/11/obj_dir --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles /home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/progress.marks
/usr/bin/gmake -f CMakeFiles/Makefile2 all
gmake[1]: Entering directory '/mnt/prj/verilator_cases/11/obj_dir'
/usr/bin/gmake -f CMakeFiles/Vt.dir/build.make CMakeFiles/Vt.dir/depend
gmake[2]: Entering directory '/mnt/prj/verilator_cases/11/obj_dir'
cd /home/yutetsu/prj/verilator_cases/11/obj_dir && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/yutetsu/prj/verilator_cases/11 /home/yutetsu/prj/verilator_cases/11 /home/yutetsu/prj/verilator_cases/11/obj_dir /home/yutetsu/prj/verilator_cases/11/obj_dir /home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/DependInfo.cmake --color=
Dependee "/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/DependInfo.cmake" is newer than depender "/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/depend.internal".
Dependee "/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/depend.internal".
Scanning dependencies of target Vt
gmake[2]: Leaving directory '/mnt/prj/verilator_cases/11/obj_dir'
/usr/bin/gmake -f CMakeFiles/Vt.dir/build.make CMakeFiles/Vt.dir/build
gmake[2]: Entering directory '/mnt/prj/verilator_cases/11/obj_dir'
[ 14%] Building CXX object CMakeFiles/Vt.dir/src/main.cpp.o
/usr/bin/c++  -DVM_COVERAGE=0 -DVM_SC=0 -DVM_TRACE=0 -DVM_TRACE_FST=0 -DVM_TRACE_VCD=0 -I/home/yutetsu/prj/verilator_cases/11/obj_dir -I/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/Vt.dir -I/home/yutetsu/prj/verilator/include -I/home/yutetsu/prj/verilator/include/vltstd   -o CMakeFiles/Vt.dir/src/main.cpp.o -c /home/yutetsu/prj/verilator_cases/11/src/main.cpp
[ 28%] Building CXX object CMakeFiles/Vt.dir/CMakeFiles/Vt.dir/Vt.dir/Vt.cpp.o
/usr/bin/c++  -DVM_COVERAGE=0 -DVM_SC=0 -DVM_TRACE=0 -DVM_TRACE_FST=0 -DVM_TRACE_VCD=0 -I/home/yutetsu/prj/verilator_cases/11/obj_dir -I/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/Vt.dir -I/home/yutetsu/prj/verilator/include -I/home/yutetsu/prj/verilator/include/vltstd   -o CMakeFiles/Vt.dir/CMakeFiles/Vt.dir/Vt.dir/Vt.cpp.o -c /home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/Vt.dir/Vt.cpp
[ 42%] Building CXX object CMakeFiles/Vt.dir/CMakeFiles/Vt.dir/Vt.dir/Vt__Syms.cpp.o
/usr/bin/c++  -DVM_COVERAGE=0 -DVM_SC=0 -DVM_TRACE=0 -DVM_TRACE_FST=0 -DVM_TRACE_VCD=0 -I/home/yutetsu/prj/verilator_cases/11/obj_dir -I/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/Vt.dir -I/home/yutetsu/prj/verilator/include -I/home/yutetsu/prj/verilator/include/vltstd   -o CMakeFiles/Vt.dir/CMakeFiles/Vt.dir/Vt.dir/Vt__Syms.cpp.o -c /home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/Vt.dir/Vt__Syms.cpp
[ 57%] Building CXX object CMakeFiles/Vt.dir/home/yutetsu/prj/verilator/include/verilated.cpp.o
/usr/bin/c++  -DVM_COVERAGE=0 -DVM_SC=0 -DVM_TRACE=0 -DVM_TRACE_FST=0 -DVM_TRACE_VCD=0 -I/home/yutetsu/prj/verilator_cases/11/obj_dir -I/home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles/Vt.dir/Vt.dir -I/home/yutetsu/prj/verilator/include -I/home/yutetsu/prj/verilator/include/vltstd   -o CMakeFiles/Vt.dir/home/yutetsu/prj/verilator/include/verilated.cpp.o -c /home/yutetsu/prj/verilator/include/verilated.cpp
[ 71%] Linking CXX executable Vt
/usr/bin/cmake -E cmake_link_script CMakeFiles/Vt.dir/link.txt --verbose=1
/usr/bin/c++     CMakeFiles/Vt.dir/src/main.cpp.o CMakeFiles/Vt.dir/CMakeFiles/Vt.dir/Vt.dir/Vt.cpp.o CMakeFiles/Vt.dir/CMakeFiles/Vt.dir/Vt.dir/Vt__Syms.cpp.o CMakeFiles/Vt.dir/home/yutetsu/prj/verilator/include/verilated.cpp.o  -o Vt 
gmake[2]: Leaving directory '/mnt/prj/verilator_cases/11/obj_dir'
[100%] Built target Vt
gmake[1]: Leaving directory '/mnt/prj/verilator_cases/11/obj_dir'
/usr/bin/cmake -E cmake_progress_start /home/yutetsu/prj/verilator_cases/11/obj_dir/CMakeFiles 0
```

Reply:
Ok, makes sense then, thanks for the experiment.